File: src/python/algorithms/array/gasStation/gas_station.py
# Greedy  TC: O(n) SC: O(1)
def can_complete_circuit(gas, cost):
    if sum(gas) < sum(cost):
        return -1

    total = 0
    res = 0

    for i in range(len(gas)):
        total += gas[i] - cost[i]

        if total < 0:
            total = 0
            res = i + 1

    return res


if __name__ == "__main__":
    test_cases = [
        ([1, 2, 3, 4, 5], [3, 4, 5, 1, 2], 3),
        ([2, 3, 4],        [3, 4, 3],       -1),
        ([5, 1, 2, 3, 4],  [4, 4, 1, 5, 1], 4),
        ([3, 3, 4],        [3, 4, 4],       -1),
        ([1, 2],           [2, 1],           1),
    ]

    for gas, cost, expected in test_cases:
        result = can_complete_circuit(gas, cost)
        status = "PASS" if result == expected else "FAIL"
        print(f"[{status}] result={result} expected={expected}")

File: src/javascript/algorithms/array/gasStation/gasStation.js
// Greedy  TC: O(n) SC: O(1)
function canCompleteCircuit(gas, cost) {
    const totalGas = gas.reduce((a, b) => a + b, 0);
    const totalCost = cost.reduce((a, b) => a + b, 0);

    if (totalGas < totalCost) return -1;

    let total = 0;
    let res = 0;

    for (let i = 0; i < gas.length; i++) {
        total += gas[i] - cost[i];

        if (total < 0) {
            total = 0;
            res = i + 1;
        }
    }

    return res;
}

function runTests() {
    const tests = [
        { gas: [1, 2, 3, 4, 5], cost: [3, 4, 5, 1, 2], expected: 3 },
        { gas: [2, 3, 4],        cost: [3, 4, 3],        expected: -1 },
        { gas: [5, 1, 2, 3, 4],  cost: [4, 4, 1, 5, 1],  expected: 4 },
        { gas: [3, 3, 4],        cost: [3, 4, 4],         expected: -1 },
        { gas: [1, 2],           cost: [2, 1],             expected: 1 },
    ];

    tests.forEach(({ gas, cost, expected }) => {
        const result = canCompleteCircuit(gas, cost);
        const status = result === expected ? "PASS" : "FAIL";
        console.log(`[${status}] result=${result} expected=${expected}`);
    });
}

runTests();
